### PR TITLE
Only create a single BoundaryCondition object per id

### PR DESCRIPTION
**Pull request by aeslaughter**
(refs #1022)

### CODE REVIEW

File: framework/src/base/NonlinearSystem.C [hunk 1]
```diff
@@ -510,44 +510,40 @@ NonlinearSystem::addScalarKernel(const  std::string & kernel_name, const std::st
 void
 NonlinearSystem::addBoundaryCondition(const std::string & bc_name, const std::string & name, InputParameters parameters)
 {
-  std::vector<BoundaryName> boundaries = parameters.get<std::vector<BoundaryName> >("boundary");
+  // Create a set of boundary ids
+  std::set<BoundaryID> boundary_ids;
+  const std::vector<BoundaryName> & boundary_names = parameters.get<std::vector<BoundaryName> >("boundary");
+  for (std::vector<BoundaryName>::const_iterator it = boundary_names.begin(); it != boundary_names.end(); ++it)
+    boundary_ids.insert(_mesh.getBoundaryID(*it));
```
Comment (permcody) [inline framework/src/base/NonlinearSystem.C]: You can construct a set from a vector in one step, it's a built-in constructor
```
s(v.begin(), v.end());
```

File: framework/src/base/NonlinearSystem.C [hunk 2]
```diff
@@ -510,44 +510,40 @@ NonlinearSystem::addScalarKernel(const  std::string & kernel_name, const std::st
...
+    {
+      _bcs[tid].addNodalBC(boundary_ids, nbc);
+      _vars[tid].addBoundaryVars(boundary_ids, nbc->getCoupledVars());
     }
+    else if (ibc.get())
+    {
+      _bcs[tid].addBC(boundary_ids, ibc);
+      _vars[tid].addBoundaryVars(boundary_ids, ibc->getCoupledVars());
+    }
+    else
```
Comment (permcody) [inline framework/src/base/NonlinearSystem.C]: Ugh, I hate this code - I know you didn't write it but it's like anti-polymorphism.  Take an inherited type, dynamically cast it, then run it through a cascading if statement to call a method...

File: framework/src/base/NonlinearSystem.C (hunk 1 above)
Comment (aeslaughter) [inline framework/src/base/NonlinearSystem.C]: Yes, but we are given names and we need the IDs, 517 calls getBoundaryID.

File: framework/src/base/NonlinearSystem.C (hunk 2 above)
Comment (aeslaughter) [inline framework/src/base/NonlinearSystem.C]: I agree, we have this same sort of thing with UserObjects.

File: framework/src/base/NonlinearSystem.C (hunk 1 above)
Comment (permcody) [inline framework/src/base/NonlinearSystem.C]: ok

**Pull request merged**